File: frontend/src/app/services/item.service.ts
import {inject, Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {CreateItemModel, ItemModel} from '../models/item.model';
import {CategorizedItemsModel} from "../models/category.model";

@Injectable({
  providedIn: 'root'
})
export class ItemService {

  private readonly http = inject(HttpClient);

  getItemsForList(id: number): Observable<CategorizedItemsModel[]> {
    return this.http.get<CategorizedItemsModel[]>(`/lists/${id}/categories/items`);
  }

  updateItem(listId: number, item: ItemModel): Observable<ItemModel> {
    return this.http.put<ItemModel>(`/lists/${listId}/items/${item.id}`, item);
  }

  createItem(listId: number, categoryId: number, item: CreateItemModel): Observable<ItemModel> {
    return this.http.post<ItemModel>(`/lists/${listId}/categories/${categoryId}/items`, item);
  }

  deleteItem(listId: number, id: number): Observable<void> {
    return this.http.delete<void>(`/lists/${listId}/items/${id}`);
  }
}

File: backend/src/main/java/fr/jbellion/shoppinglist/service/ItemService.java
package fr.jbellion.shoppinglist.service;

import fr.jbellion.shoppinglist.dto.CategorizedItemsDto;
import fr.jbellion.shoppinglist.dto.item.ItemDto;

import java.util.List;

public interface ItemService {

  List<CategorizedItemsDto> getItemsListByCategories(long listId);

  ItemDto updateItem(long listId, ItemDto item, Long originClientId);

  void deleteItem(long listId, long id, Long originClientId);

  ItemDto createItem(long listId, long categoryId, ItemDto item, Long originClientId);
}
